internal/engine/wazevo/engine.go

```diff
@@ -227,27 +226,81 @@ func (e *engine) compileModule(ctx context.Context, module *wasm.Module, listene
 	needCallTrampoline := callTrampolineIslandSize > 0
 	var callTrampolineIslandOffsets []int // Holds the offsets of trampoline islands.
 
-	for i := range module.CodeSection {
-		if wazevoapi.DeterministicCompilationVerifierEnabled {
-			i = wazevoapi.DeterministicCompilationVerifierGetRandomizedLocalFunctionIndex(ctx, i)
-		}
+	type CompiledLocalFuncResult struct {
+		Body             []byte
+		RelsPerFunc      []backend.RelocationInfo
+		IDX              wasm.Index
+		SourceOffsetInfo []backend.SourceOffsetInfo
+	}
+
+	compiledFuncs := make([]CompiledLocalFuncResult, len(module.CodeSection))
+
+	workers := runtime.GOMAXPROCS(0)
+
+	wg := sync.WaitGroup{}
+	wg.Add(workers)
+
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
+
+	sections := sequence(len(module.CodeSection))
```

Comment:
based on the errors you guys mentioned, should be resolved with:
1. create a seeded PRNG.
2. shuffle the code sections using the PRNG.
3. create a thread safe queue to pop the module function to compile and its idx after shuffle
4. guard the compiledFuncs with a lock. and insert the result into the idx.

Reply:
(this approach should allow removing the conditional logic on deterministic builds across the code base as well while enabling PIE)

Reply:
I am not sure I follow... especially the bit about guarding the compiled funcs with a lock. We want them to run concurrently as this is how we are getting our compilation performance gains. 

Do you mean only for the deterministic checking mode?

Reply:
Thanks! I look forward to it. I knew I was not following :)

Reply:
https://github.com/tetratelabs/wazero/pull/2394 take a peek. just opened a PR to run the ci/cd to see if it works out. feel free to merge the changes into this PR and close it if it works out.

Reply:
mm I feel like what we really want is to make the build independent of the order; compiling each function should not really depend on shared state, if it does, then something icky must be going on :p

Reply:
agreed, the fix i opened that PR was to resolve the deterministic builds error you mentioned, not to ensure a particular order of compilation. by shuffling we ensure the compilation *is* randomized but can be deterministically reproduced by controlling the seed.

interestingly the errors from CI have changed dramatically from my changes, they likely identify any shared state now that the compiled functions are truly randomized.